### ISSUE TITLE
Should our 1-element array literals require a trailing comma?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#21092]
A one-tuple in Chapel is represented as `(elem, )` since simply `(elem)` is ambiguous whether we mean a 1-tuple or simply a parenthesized expression.  Meanwhile, for one-element array literals, we support either `[elem]` or `[elem, ]`.  I believe the reason for this is that there isn't the same sort of ambiguity.

[This SO question](https://stackoverflow.com/questions/74575947/chapel-ranges-defined-using-bounds-of-type-rangeint64-bounded-false-are) makes me wonder whether this is a mistake, since even though it is unambiguous, it could cause someone unfamiliar with the language to get a surprising result due to misplaced brackets.  Whereas if the compiler gave them an error stating that `[expr]` is not a legal expression, and that if their goal was to get a one-element array literal they should add a trailing comma, then such mistakes would be less problematic.  In saying this, I'm assuming that 1-element arrays are not likely to be all that common.

Making this change would also make 1-element arrays more consistent with 1-tuples, which seems mildly attractive.

**Comment by lydia-duncan** [comment]
I would want the error message to be specific about where the incorrect `[expr]` is used, since I don't think we intend to make `A[expr]` invalid.  But I otherwise don't have a major problem with this suggestion

**Comment by bradcray** [comment]
@lydia-duncan : That would be a different syntactic rule in the parser (the one representing indexing expressions rather than array literals).  Specifically, I'm imagining we would take the production at:  https://github.com/chapel-lang/chapel/blob/main/frontend/lib/parsing/chpl.ypp#L3773-L3776 and turn it into an error case rather than a "build array literal" case.

**Comment by bradcray** [comment]
@mppf: I'm curious whether you have an opinion here since you implemented the support for trailing commas in tuples and array literals in https://github.com/chapel-lang/chapel/pull/4521 and I believe were a champion (or ambassador) for the trailing comma approach to begin with.

Meanwhile, I'm going to try making that rule an error to see whether we have any key tests or code that rely upon it and make me change my mind here (I'm not expecting that).

**Comment by lydia-duncan** [comment]
> That would be a different syntactic rule in the parser (the one representing indexing expressions rather than array literals). Specifically, I'm imagining we would take the production at: https://github.com/chapel-lang/chapel/blob/main/frontend/lib/parsing/chpl.ypp#L3773-L3776 and turn it into an error case rather than a "build array literal" case.

My point was about users not the parser.  I think we want to phrase the error message so that it only talks about array literals or creating an array, so that it is clear `A[expr]` is not impacted.

**Comment by bradcray** [comment]
Oh, I didn't understand that.  So you're saying that if a line contained both a 1-element array literal and an indexing expression, like:

```chapel
var B = [A[i]];
```

you'd want to make sure the error made it clear that the problem was with `[<expr>]` not being `[<expr>, ]` rather than with `A[i]`?  (e.g., by giving a column number for the error or printing out the expression in question in some way?).  Would you give an example of what you'd like the error to say for this example given how close the expressions are to one another?

I agree that such an error would be ideal, but wouldn't insist on it before implementing some form of deprecation in order to get users to start using this prior to 2.0.  Specifically, I think that 1-element arrays are generally quite rare, and ones that lines of code that mix these two forms are even rarer.  That said, maybe this would be easy to do with the new error framework—I'll ask the experts (Daniel or Anna) once I have more confidence we should make this change.

**Comment by mppf** [comment]
> @mppf: I'm curious whether you have an opinion here since you implemented the support for trailing commas in tuples and array literals in #4521 and I believe were a champion (or ambassador) for the trailing comma approach to begin with.

I think it's reasonable to require the comma for 1-element array literals for consistency with tuples & to make errors more understandable. I'm not aware of a particularly compelling counter-argument (in particular, as far as I know, 1-element array literals aren't particularly common, as you said). However I could go along with leaving it the way it is now, if we find some reason to do that.

**Comment by lydia-duncan** [comment]
I think just something like "creating a single element array requires a ',' after the element" would be sufficient for me and I don't think that would require any special handling.  My worry is specifically around `[expr]` versus the words "creating a single element array" - the latter seems less likely to result in misunderstanding due to its specificity about the situation involved.  It had seemed like you were leaning towards using the square brackets, which is why I spoke up

**Comment by vasslitvinov** [comment]
I can go either way on this proposal.

What I want to discuss is the brackets in the SO question as a motivation. Granted, I could have a similar reaction to that question, wanting to disallow that kind of scenario. OTOH how far should we / can we go? For example, another user could post a question where they wrote:
```chpl
forall i in (1..n,1..n) {
  ....
}
```
intending a parallel iteration across a 2-D space. Would that be a motivation for us to do something about tuples of ranges? I suspect not. The question with brackets looks analogous w.r.t. how little it motivates me to do this change.

**Comment by bradcray** [comment]
> Would that be a motivation for us to do something about tuples of ranges? I suspect not.

I agree, and also agree with your point that we can't protect all users from all problems that might create for themselves.  But given that 1-element array literals are (hypothesized to be—I'm checking as we speak) very rare, not particularly useful, and represent a case where the same thing can be written two different ways with  no impact, supporting both doesn't seem to be very essential.  I think the issue is also exacerbated a bit by the fact that square brackets are so overloaded in Chapel.  Even looking at the user's code in SO, I found myself wondering whether the issue was that the `[1..n]`-type expression was being interpreted as a body-less forall loop (like `[_ in 1..n]` or `[_ in 1..n] ;`).

So basically, I don't think we would break 2.0 over this if we'd run into it a year from now.  But it seems like an attractive unification and improvement prior to the 2.0 freeze to me (again, barring compelling cases in testing or user code that change our minds).

**Comment by bradcray** [comment]
I'm still a supporter of this proposal (requiring array literals to have a trailing comma), but ran into a case that gave me slight pause yesterday, so wanted to note it here.  The argparse library takes in lots of its arguments as arrays, and use cases in our testing and mason contain code like:

```chapel
   var vcsFlag = parser.addFlag(name="vcs",
                              opts=["--no-vcs"],
                                defaultValue=false);
```

that would need to change to have a trailing comma for such 1-element `opts` arguments:

```chapel
...
                              opts=["--no-vcs", ],
```

This use case (passing 1-element arrays in as arguments) isn't one I'd considered when suggesting that 1-element tuples were unusual in practice, and it definitely seems slightly annoying or dissatisfying in this context compared to not having to type the comma.  At the same time, I continue to think that square brackets are sufficiently overloaded in Chapel that taking the stricter approach will probably head off more errors and surprises than it will cause pain, and think it's rationalizable in order to get the consistency with tuples (specifically, my next thought was "Maybe tuples would be better suited for these types of multi-value arguments as well" only to realize that they'd also require the trailing comma.

Tagging @arezaii as the argparse author (as well as mason's use of it) for comments/concerns in case he's not tracking this issue.

**Comment by bradcray** [comment]
The Spawn module's interface and tests also involve such cases, so @mppf and @daviditen may want to consider that as well.  E.g.,

```diff
-var sub = spawn(["cat"], stdin=pipeStyle.bufferAll, stdout=pipeStyle.pipe);
+var sub = spawn(["cat", ], stdin=pipeStyle.bufferAll, stdout=pipeStyle.pipe);
```

**Comment by bradcray** [comment]
Oh, and the HDF5 module, @daviditen:

```diff
-          var inOffsetArr = [inOffset: C_HDF5.hsize_t],
-              inCountArr  = [readCount: C_HDF5.hsize_t];
+          var inOffsetArr = [inOffset: C_HDF5.hsize_t, ],
+              inCountArr  = [readCount: C_HDF5.hsize_t, ];
 
           C_HDF5.H5Sselect_hyperslab(dataspace,
                                      C_HDF5.H5S_SELECT_SET,
                                      c_ptrTo(inOffsetArr), nil,
                                      c_ptrTo(inCountArr), nil);
```

**Comment by vasslitvinov** [comment]
> The Spawn module's interface and tests also involve such cases, so @mppf and @daviditen may want to consider that as well.

This is more an argument for allowing `spawn()` to accept just a string for its first argument. "No brainer" to me. Yes, there is the 'executable' argument and 'spawnshell', however those are different.

**Comment by daviditen** [comment]
Requiring the trailing comma seems fine to me in the Spawn and HDF5 cases. It is a bit concerning though that this will be an interface change to Spawn which has already been declared 2.0 complete.

Will this include deprecating single element array literals without the comma for a few releases?

**Comment by arezaii** [comment]
I think argparse will be OK if the syntax changes to require the trailing comma. I'm not a huge fan of the trailing comma in any case because it just seems odd (it makes me think there's an element that was intended, but missing/forgotten). If a required trailing comma is what we need to disambiguate the various overloads of `[ ]` and align with single-element tuples then I am happy to make the changes to argparse tests and the Mason implementation when needed.

**Comment by bradcray** [comment]
@daviditen : 

> Will this include deprecating single element array literals without the comma for a few releases?

Yes.

> It is a bit concerning though that this will be an interface change to Spawn which has already been declared 2.0 complete.

I don't think this is an interface change—rather it's a change to how one-element array literals that might be passed to the stabilized interface are written.  What am I missing?

(even if we were to adopt Vass's proposal, which I agree with Michael should be a separate discussion/issue if anyone wants to champion it, that's not a breaking change, since 2.0 doesn't prevent new capabilities from being added to a stabilized module).

**Comment by daviditen** [comment]
> I don't think this is an interface change—rather it's a change to how one-element array literals that might be passed to the stabilized interface are written. What am I missing?

Breaking existing calls feels like an interface change, but I suppose you're right that technically the interface to the functions hasn't changed.

**Comment by bradcray** [comment]
@arezaii : 

> I am happy to make the changes to argparse tests and the Mason implementation when needed.

I've already made the required changes to mason and argparse (and tests), so if we proceed with the proposal in this issue, no additional work is needed there (but thanks for the offer!)

> If a required trailing comma is what we need to disambiguate the various overloads of [ ] and align with single-element tuples

I don't believe it's required for the language to disambiguate, but proposed it (on this issue and [on Discourse](https://chapel.discourse.group/t/design-one-element-array-syntax/17680)) to attempt to reduce the potential for user confusion and mistakes (like the SO question linked in the OP).  Since the response had been strictly positive (until now), I was proceeding with trying to implement it, in part to see whether any worrisome cases came up in testing.  My last few comments were trying to highlight such cases to see if they gave anyone additional pause.  It sounds like you object more to the original proposal, though?  This would be a good time to make your stand if you want to oppose the proposed direction.

(I'll note that an advantage to making the change now in the face of uncertainty is that if we end up deciding in the long-term that it's too annoying, we can re-introduce `[elem]` as a single-element array literal; whereas if we proceed with things as they are on `main` today, and continue to have users being confused, we can't change it prior to Chapel 3.0).

**Comment by arezaii** [comment]
To be sure, I am only mildly annoyed by the existence of the trailing comma, certainly not enough to derail the proposal. Since it brings alignment with the single-element tuple, maybe it will grow on me 😃

**Comment by bradcray** [comment]
It also occurs to me:  Rather than deprecating comma-less syntax, we could simply support it for all time, but have an "on-by-default", yet squashable, warning for it which would potentially achieve the same result as the proposal in the OP by helping users realize they had written an array literal unwittingly.  I think it'd be the first case of a purely stylistic warning, which makes me somewhat reluctant about it; and users who preferred that style would need to squash the warning, which might seem annoying or make the comma-less form feel second-class.  But I wanted to throw it out as a possibility (nearly equivalently, we could never remove the deprecation warning).

**Comment by vasslitvinov** [comment]
It looks to me like this proposal is aimed to help users who are less familiar/fluent with Chapel at the expense of those who are more experienced.

I do not know how many users have bumped into this issue and how significant of a hurdle it was for them, so I am not judging whether this is a good trade-off. However, what if instead we beef up the compiler to watch out for this pattern and give suggestions to the user when it occurs?

**Comment by bradcray** [comment]
> at the expense of those who are more experienced.

Well... arguably if our rule had always been "one-element arrays require trailing commas" then those experienced users still wouldn't have a problem because they'd be experienced with the rule.  :)  Unless by "expense" you mean "having to type the extra comma at all" rather than "having to adhere to a new rule."

> to watch out for this pattern

What pattern are you imagining we would look for?

**Comment by bradcray** [comment]
Follow-up question:

Should 1-element associative array literals `["Brad"=>4]` also require a trailing comma?

:+1: = yes, because consistency!
:-1: = no, because `=>` is sufficiently weird and different that nobody's going to accidentally write this if not making an associative array literal!

**Comment by arezaii** [comment]
if 👍 =1 and 👎 =7, I'm about a 4.

**Comment by bradcray** [comment]
w.r.t. the overall change, an internal poll strongly favored continuing to make the change for the rectangular array literal case, so I'm planning to proceed with that.

A second poll shows a nearly-as-strong desire to make the same change for associative array literals.  While I could go either way on that one, the poll makes me inclined to make the change:
* because it's consistent
* because I think one-element associative array literals are far less common than rectangular array literals

In both cases, a point to take comfort in is that we could always relax these decisions post-2.0 without breaking code if we decided they were wrong.  But we can't make these changes post-2.0 without breaking things.

**Comment by bradcray** [comment]
Sooo....  While I postulated above that "nobody really uses one-element array literals anyway, do they?" I apparently never actually checked the CHAMPS code, which does in spades (~80 such cases).  I've reached out to them to get their thoughts, but the experience of updating them has definitely given me pause (pause along the lines of "Would we just want to warn for specific cases that seem like potential misunderstandings, like "array literals of 1-element containing range literals?"  This is the sort of thing I imagine Vass may have been suggesting [above](https://github.com/chapel-lang/chapel/issues/21092#issuecomment-1376582380).

**Comment by bradcray** [comment]
This got closed due to a rebase / force push, so re-re-opening it.

**Comment by bradcray** [comment]
My current thinking here is to disable the "require trailing comma on 1-element array literals" warning for 1.30, potentially replacing it with a new warning to head off the original error (thinking that `for i in [1..n]` would result in i taking on the integers 1, 2, 3, ...n).  Some options I'm weighing include:

* warning for 1-element array literals of range-literals (i.e., `[1..n]` or `[1..n,]`)
* warning for 1-element array literals of range-literals, only when there's no trailing comma (i.e., `[1..n]`)
* warning for any array literal of range-literals (i.e., `[1..n]`, `[1..n,]`, `[1..n, 1..n]`, `[1..n, 1..n,]`, etc.)
* warning for any array literal of range-literals, when there's no trailing comma (i.e., `[1..n]`, `[1..n, 1..n]`, etc.)
* no warning at all

I think I'm currently leaning toward the second bullet, primarily because the last warning blew up in my face so much, and the trailing comma provides a way to say "I know what I'm doing" without having to throw a compiler flag.  But I'm interested in other opinions here.

**Comment by mppf** [comment]
FWIW, I also like the 2nd option. I like that it provides an easy way to resolve the warning. I'm tempted to aim further, for the 4th option (to include multiple range literals) but I'm less sure of the implications there and it's further from the original problematic case. (In particular, if somebody thought `for i in [1..n]` would loop over the integers 1 to n; I can imagine that they might also think that `for (i,j) in [1..n, 1..n]` would iterate over the 2D region... But that case already results in an error).

Details about the 2D case below -- but note, that error probably needs improvement if it's likely a new user will face it.

<details>

``` chapel
config const n = 10;
for (i,j) in [1..n, 1..n] {
  writeln(i, j);
}
```

```
% chpl aa.chpl
aa.chpl:2: error: illegal tuple variable declaration with non-tuple initializer
```

</details>

**Comment by bradcray** [comment]
Note that if a user mistook `[1..n, 1...n]` for a 2D index space and was comfortable computing with tuples, they could have written `for ij in [1..n, 1..n]` and be caught in a similar bind to `for i in [1..n]`.  Let me know whether that changes your opinion.

**Comment by mppf** [comment]
FWIW, it does not change my opinion, because I consider `for ij in [1..n, 1..n]` "more advanced" than `for (i,j) in [1..n, 1..n]`; and so less likely a stumbling block for a new user. However, I am certainly open to going with the 4th option if that ends up being what you prefer. In this 2D case here as well as in general, we can add warnings in the future if/when users run into confusion with specific patterns.
